lib/ddtrace/contrib/httprb/patcher.rb

```diff
@@ -0,0 +1,38 @@
+require 'ddtrace/contrib/patcher'
+require 'ddtrace/contrib/httprb/instrumentation'
+require 'ddtrace/contrib/httprb/datadog_wrap'
+require 'http'
+
+module Datadog
+  module Contrib
+    # Datadog Httprb integration.
+    module Httprb
+      # Patcher enables patching of 'httprb' module.
+      module Patcher
+        include Contrib::Patcher
+
+        module_function
+
+        def patched?
+          done?(:httprb)
+        end
+
+        # patch applies our patch
+        def patch
+          do_once(:httprb) do
+            begin
+              register_feature!
+              ::HTTP::Options.send(:include, Instrumentation)
```

Comment:
Why not use the native [`ActiveSupport::Notifications` events](https://github.com/httprb/http/blob/f311a9b5d27042f581b4073c70fb5d2763f31a27/lib/http/features/instrumentation.rb#L32-L38)?

They have an event for the request (`start_request.http`) and one for the response (`request.http`). You'll still have to stitch them together like you are already doing, but the integration will require no monkey-patching.

Reply:
my understanding from some very brief testing of this is that even when using the active support notification events you still have to enable the `use(instrumentation: <instrumentation>)` feature as documented here https://github.com/httprb/http/wiki/Logging-and-Instrumentation#instrumentation

So monkey patching of `HTTP::Options` would still be required afaik, additionally I was concerned this approach would interfere/overwrite/enter into a race condition/etc with any other attempts by users to include the `use` `logging` or `instrumentation` features (I still have to test this) and just I figured i'd remove this possibility by going one layer lower in the stack.

I will test this out as soon as I get some cycles and document, happy to refactor this up a layer to use the canonical `instrumentation` feature if it allows multiple instrumenters and still permits http header injection for distributed tracing. Either way I do think we'll need to keep some monkey patching in there.

Reply:
I clean this up a good amount to be something close to reviewable and also tested out the behavior we discussed here re: monkey patching and active support notifications. You can find a demonstration of these details [here](https://gist.github.com/ericmustin/69dbbab70f4a74ad60b06c6602429422) (requires `activesupport` and `http` gems locally) :

https://gist.github.com/ericmustin/69dbbab70f4a74ad60b06c6602429422

As you can see by running the above `$ ruby test.rb` I get the following output

```
COMP10929:httprb_testing eric.mustin$ ruby test_app.rb 
we should see our instrumentation occur but only the last instrumentation passed
starting dd instrumentation
ending dd instrumentation
200
we won't see instrumentation occur because we haven't explicitly called it
200
```

Essentially, if we attempt to modify the existing `instrumentation` feature, due to the way the `features` method is implemented [here](https://github.com/httprb/http/blob/6e4168f04ff5130647efb26a63f8f445ba4a6de9/lib/http/options.rb#L106-L117), we simply overwrite the last instrumentation value assigned. So what would happen is that in any situation where the user is currently calling `HTTP` with something like `HTTP.use(instrumentation: <instrumentation_feature_details>)` in their application code, our instrumentation would get lost. Additionally, as mentioned previously and demonstrated above, the `ActiveSupport::Notifications` events are only generated when the user explicitly calls `.use(instrumentation: <instrumentation_feature_details>)`. There's no real global  setting to enable a feature, only to register a new one. So i do think both adding  the custom feature (so as not to interfere with any existing instrumentation) and monkey patching to inject the feature as a default option, are appropriate here. 

Let me know what you think, it's possible i've missed something.